Sources/Testing/CMakeLists.txt

```diff
@@ -0,0 +1,92 @@
+# This source file is part of the Swift.org open source project
+#
+# Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+# Licensed under Apache License v2.0 with Runtime Library Exception
+#
+# See http://swift.org/LICENSE.txt for license information
+# See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+
+add_library(Testing
+  EntryPoints/ABIEntryPoint.swift
+  EntryPoints/EntryPoint.swift
+  EntryPoints/SwiftPMEntryPoint.swift
+  EntryPoints/XCTestScaffold.swift
+  Events/Clock.swift
+  Events/Event.swift
+  Events/Recorder/Event.ConsoleOutputRecorder.swift
+  Events/Recorder/Event.HumanReadableOutputRecorder.swift
+  Events/Recorder/Event.JUnitXMLRecorder.swift
+  Events/Recorder/Event.Symbol.swift
+  Events/TimeValue.swift
+  ExitTests/ExitCondition.swift
+  ExitTests/ExitTest.swift
+  ExitTests/WaitFor.swift
+  Expectations/Expectation.swift
+  Expectations/Expectation+Macro.swift
+  Expectations/ExpectationChecking+Macro.swift
+  Issues/Confirmation.swift
+  Issues/ErrorSnapshot.swift
+  Issues/Issue.swift
+  Issues/Issue+Recording.swift
+  Issues/KnownIssue.swift
+  Parameterization/CustomTestArgumentEncodable.swift
+  Parameterization/Test.Case.Generator.swift
+  Parameterization/Test.Case.ID.swift
+  Parameterization/Test.Case.swift
+  Parameterization/TypeInfo.swift
+  Running/Configuration.swift
+  Running/Configuration.TestFilter.swift
+  Running/Configuration+EventHandling.swift
+  Running/Runner.Plan.swift
+  Running/Runner.Plan+Dumping.swift
+  Running/Runner.RuntimeState.swift
+  Running/Runner.swift
+  Running/SkipInfo.swift
+  SourceAttribution/Backtrace.swift
+  SourceAttribution/CustomTestStringConvertible.swift
+  SourceAttribution/Expression.swift
+  SourceAttribution/Expression+Macro.swift
+  SourceAttribution/SourceContext.swift
+  SourceAttribution/SourceLocation.swift
+  Support/Additions/ArrayAdditions.swift
+  Support/Additions/CollectionDifferenceAdditions.swift
+  Support/Additions/CommandLineAdditions.swift
+  Support/Additions/NumericAdditions.swift
+  Support/Additions/ResultAdditions.swift
+  Support/CartesianProduct.swift
+  Support/CError.swift
+  Support/Environment.swift
+  Support/FileHandle.swift
+  Support/Graph.swift
+  Support/JSON.swift
+  Support/Locked.swift
+  Support/SystemError.swift
+  Support/UncheckedSendable.swift
+  Support/Versions.swift
+  Test.ID.Selection.swift
+  Test.ID.swift
+  Test.swift
+  Test+Discovery.swift
+  Test+Macro.swift
+  Traits/Bug.swift
+  Traits/Comment.swift
+  Traits/Comment+Macro.swift
+  Traits/ConditionTrait.swift
+  Traits/ConditionTrait+Macro.swift
+  Traits/HiddenTrait.swift
+  Traits/ParallelizationTrait.swift
+  Traits/SPIAwareTrait.swift
+  Traits/Tags/Tag.Color.swift
+  Traits/Tags/Tag.Color+Loading.swift
+  Traits/Tags/Tag.List.swift
+  Traits/Tags/Tag.swift
+  Traits/Tags/Tag+Macro.swift
+  Traits/Tags/Tag+Predefined.swift
+  Traits/TimeLimitTrait.swift
+  Traits/Trait.swift)
+target_link_libraries(Testing PUBLIC
```

Comment:
TestingInternals is not meant to be exposed publicly. Is this the right specification?

Reply:
(I genuinely don't know. I don't know CMake too goods.)

Reply:
In this case it's saying the fact that `Testing` links to `TestingInternal` is `PUBLIC`.  Ironically, this makes it so consumers don't have to explicitly list `TestingInternal` (e.g. [here](https://github.com/thebrowsercompany/swift-webdriver/blob/ffd8269d771dc1e33098305dcf58c94859bbf04a/Tests/CMakeLists.txt#L13)) as a link target, since CMake will recognize the transitive relationship via this definition, keeping it nicely contained.

Reply:
Just to be clear: will it be possible for code outside this repository to say `import TestingInternals`? (We don't want that.)

Reply:
Indeed, it is critical we don't leak it outward because it includes redefinitions of symbols from the platform's C standard library that will make linking a nightmare if they escape.

Reply:
When I mark the linkage here as `PRIVATE`, when I try to build the `swift-webdriver` test target linked to `Testing` I get: 
```
X:\Tests\UnitTests\APITests.swift:2:8: error: missing required module 'TestingInternals'
  1 | import Foundation
  2 | import Testing
    |        `- error: missing required module 'TestingInternals'
  3 | import TestsCommon
  4 | @testable import WebDriver
```

I can add this to the `SwiftWebDriverPackageTests` executable's `target_link_libraries()`, but it seems best to try to avoid that, which is why I made this `PUBLIC` initially.  

I'm not sure how to link to `Testing` without somehow telling CMake where to find `TestingInternals`, so I'm not sure how hermetic this can be.  I'm curious how SwiftPM does this?

Reply:
swift-corelibs-foundation uses CMake and links to several subcomponents as `PRIVATE`. Worth looking at what's being done there?

Reply:
See also DispatchStubs in swift-corelibs-dispatch.

Reply:
Those do not use the ACL but rather the `@_implementationOnly` which differs in that it does not serialise the dependency out so the module search path is not required for consumers of the module.

Reply:
AFAIK `private` imports do not work with CMake's `PRIVATE` for non-resilient modules, only `@_implementationOnly` does. This is why https://github.com/apple/swift-package-manager/pull/7026 kept using `@_implementationOnly` instead of switching to `private`.

Reply:
I'll check it out.

Reply:
Verified this works.  AFAIK CMake doesn't define anything, so I went with your suggestion which seems clear.

Reply:
Thanks for taking a look! For comparison, SPM uses `USE_IMPL_ONLY_IMPORTS` for the same purpose. So I think we're in good company here.

Reply:
And swift-syntax uses `SWIFT_SYNTAX_BUILD_USING_CMAKE`. :)

Reply:
@neonichu - could you share how SPM works with this? Is it a matter of flags? If so, we should be able to get the right behaviour with CMake as well. If not, this feels like it should be predicated on Darwin vs non-Darwin.

Reply:
I haven't looked into it deeply, but I think it's actually because CMake's `PRIVATE` works more correctly by not making the module interface available to clients vs. SwiftPM dumping all modules into a single directory, so any module built from source is visible to anyone.

Reply:
Does that mean somebody adopting swift-testing via SwiftPM would be able to access TestingInternals? That would be unfortunate.

Reply:
Yes, though since it is a C-language module, it won't be accessible to every target, "just" all targets that transitively import `Testing`.

Reply:
I'll see about renaming the module so it at least won't show up in autocomplete in Xcode.

Reply:
@bnbarham suggested adding `target_compile_options(Testing PRIVATE -enable-library-evolution)` to suppress clients being able to see the internal module. That's not a bad idea in general, probably.

Reply:
Allll the swift modules.

Reply:
Yeah, enabling library evolution sounds like a good idea.

Reply:
I just landed a PR to rename `TestingInternals` to `_TestingInternals` to make it at least not show up in Xcode's autocomplete suggestions (I… hope?) Just FYI, that'll conflict with this change.

Reply:
Should I merge your changes or do a rebase -> force push?

Reply:
OK, should be g2g.